core/lib/Cardano/Script.hs

```diff
@@ -87,15 +89,59 @@ keyHashFromBytes bytes
 -- | Validate 'Script'
 --
 -- @since 3.0.0
-validateScript :: Script -> Bool
-validateScript (RequireSignatureOf _) = True
-validateScript (RequireAllOf content) =
-    not (L.null content) && L.all ((== True) . validateScript) content
-validateScript (RequireAnyOf content) =
-    not (L.null content) && L.all ((== True) . validateScript) content
-validateScript (RequireMOf m content) =
-    m > 0 && L.length content >= fromIntegral (toInteger m) &&
-    L.all ((== True) . validateScript) content
+validateScript :: Script -> Either InvalidScriptError ()
+validateScript (RequireSignatureOf (KeyHash bytes)) =
+    case keyHashFromBytes bytes of
+        Just _ -> Right ()
+        Nothing -> Left WrongKeyHash
+validateScript (RequireAllOf content) = do
+    when (L.null content) $
+        Left EmptyList
+    scanContent content
+validateScript (RequireAnyOf content) = do
+    when (L.null content) $
+        Left EmptyList
+    scanContent content
+validateScript (RequireMOf m content) = do
+    when (m == 0) $
+        Left MZero
+    when (length content < fromInteger (toInteger m) ) $
+        Left ListTooSmall
+    scanContent content
+
+scanContent :: [Script] -> Either InvalidScriptError ()
+scanContent content = do
+    let lefts = filter isLeft $ map validateScript content
+    if null lefts then do
+        let isSignature (RequireSignatureOf _) = True
+            isSignature _ = False
+        let sigs = filter isSignature content
+        if length sigs == length (L.nub sigs) then
+            Right ()
+        else
+            Left DuplicateSignatures
+    else
+        head lefts
+
+-- | Errors when handling 'Script'
+--
+-- @since 3.0.0
+data ScriptError = MalformedScript | InvalidScript InvalidScriptError
+    deriving Eq
+
+data InvalidScriptError = EmptyList | ListTooSmall | MZero | DuplicateSignatures | WrongKeyHash
+    deriving (Eq, Show)
+
+instance Show ScriptError where
```

Comment:
Do :pray: 
Not :pray: 
Manually :pray: 
Define :pray: 
Show :pray: 
Instances :pray: 

^^" ... `Show` is really for _debugging_ and showing things in a Haskelly way. To get a human-readable representation of some data-type, it is better to define an explicit function like `prettyThis` or `toStringThat` or, use a type-class that convey this idea (like `Buildable` or `Pretty`)

Reply:
There is also `displayException` if you're not against adding an `Exception` instance.

Reply:
yes ... I had first toText, but then realised we do not have it...and I am used to it. corrected https://github.com/input-output-hk/cardano-addresses/pull/79/commits/f72feec2b54adcf95983289edfae21688f14d051 btw. sometimes I think it would be good to have adrestia-base or adrestia-core repo in which we have all those intensively reusable things like encode/decode, toText, fromText, ... to name a few. Each repo would use it as dependency and there would be a lot of code cutting as a result